src/org/jetbrains/plugins/scala/compiler/ScalaHydraCompilerConfigurationPanel.scala

```diff
@@ -0,0 +1,74 @@
+  private val UnknownVersion = "unknown"
+
+  val documentAdapter = new DocumentAdapter {
+      override def textChanged(documentEvent: DocumentEvent): Unit =
+        if (!getUsername.isEmpty && !getPassword.isEmpty) downloadButton.setEnabled(true)
+        else downloadButton.setEnabled(false)
+    }
+
+  val focusListener = new FocusListener {
+    override def focusGained(e: FocusEvent) = {}
+
+    override def focusLost(e: FocusEvent) = if (!getUsername.isEmpty && !getPassword.isEmpty &&
+      (HydraCredentialsManager.getLogin != getUsername || HydraCredentialsManager.getPlainPassword != getPassword)) {
+      HydraCredentialsManager.setCredentials(getUsername, getPassword)
+      hydraVersionComboBox.setItems(Versions.loadScalaVersions(Platform.Hydra))
+    }
+  }
+
+  userTextField.addFocusListener(focusListener)
+  userTextField.getDocument.addDocumentListener(documentAdapter)
+  passwordTextField.getDocument.addDocumentListener(documentAdapter)
+  passwordTextField.addFocusListener(focusListener)
+  hydraVersionComboBox.setItems(Versions.loadScalaVersions(Platform.Hydra))
+  downloadButton.addActionListener((_: ActionEvent) => onDownload())
+
+  def selectedVersion: String = hydraVersionComboBox.getSelectedItem.asInstanceOf[String]
+
+  def onDownload() = {
+    downloadVersionWithProgress(project.scalaModules.map(module => module.sdk.compilerVersion.getOrElse(UnknownVersion)), selectedVersion)
+    settings.hydraVersion = selectedVersion
+  }
+
+  private def downloadVersionWithProgress(scalaVersions: Seq[String], hydraVersion: String): Unit = {
+    val filteredScalaVersions = scalaVersions.distinct.filterNot(_ == UnknownVersion)
+      .map(Version(_)).filter(_ >= Version("2.11")).map(_.presentation).filterNot(_ == "2.12.0")
+    val filteredScalaVersionsString = filteredScalaVersions.mkString(", ")
+    val scalaVersionsToBeDownloaded = filteredScalaVersions.filterNot(settings.artifactPaths.containsKey(_))
+    val scalaVersionsToBeDownloadedString = scalaVersionsToBeDownloaded.mkString(", ")
+    if (!scalaVersionsToBeDownloaded.isEmpty) {
+      val result = extensions.withProgressSynchronouslyTry(s"Downloading Hydra $hydraVersion for $scalaVersionsToBeDownloadedString")(downloadVersion(scalaVersionsToBeDownloaded, hydraVersion))
+      result match {
+        case Failure(exception) => {
+          Messages.showErrorDialog(contentPanel, exception.getMessage, s"Error Downloading Hydra $hydraVersion for $scalaVersionsToBeDownloadedString")
+        }
+        case Success(_) => Messages.showInfoMessage(s"Successfully downloaded Hydra $hydraVersion for $scalaVersionsToBeDownloadedString", "Download Hydra Successful")
+      }
+    } else {
+      Messages.showInfoMessage(s"Hydra $hydraVersion for $filteredScalaVersionsString is already downloaded", "Hydra version already downloaded")
+    }
+  }
+
+  private def downloadVersion(scalaVersions: Seq[String], hydraVersion: String):(((String) => Unit) => Unit) =
```

Comment:
Why not declaring this method like the following:

`private def downloadVersion(scalaVersions: Seq[String], hydraVersion: String, listener: String => Unit): Unit = ... `

What are you gaining by returning a function?

Reply:
I pass the function that is returned from downloadVersion method to withProgressSynchronouslyTry[T](title: String)(body: ((String => Unit) => T)): Try[T]

Reply:
Got it. Looks good then!